<a href="https://colab.research.google.com/github/Aditinegi97/Predicting_Credit_Risk_with_Logistic_Regression/blob/main/Predicting_Credit_Risk_with_Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing the dependences

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [5]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [6]:
# loading dataset to our pandas dataframe
credit_card_data = pd.read_csv('/content/creditcard.csv')

In [7]:
#first 5 rows of the dataset
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [8]:
credit_card_data.tail()

Time        V1        V2        V3        V4        V5        V6  \
19893  30631 -0.377215  0.973528  1.647077  0.732439  0.024728 -0.541379   
19894  30631  1.209281  0.078793  0.061820  0.593730 -0.235772 -0.448524   
19895  30632  1.286596 -1.450336  0.814530 -1.308949 -2.055209 -0.592064   
19896  30633 -0.488090  1.018448  0.670593 -0.245462  0.828347 -0.233102   
19897  30633 -2.609841  2.479357  0.763844  0.044509 -0.645716  0.762867   

             V7        V8        V9  ...       V21       V22       V23  \
19893  0.828488 -0.060740 -0.725148  ...  0.228443  0.685913 -0.107687   
19894 -0.141196  0.089236  0.411825  ... -0.302369 -0.984051  0.130401   
19895 -1.317286  0.032386 -1.720017  ...  0.040743  0.262534 -0.045112   
19896  0.662586 -0.040028 -0.279439  ... -0.344859 -0.902035 -0.050171   
19897 -1.626415 -7.617854  1.399746  ...       NaN       NaN       NaN   

            V24       V25       V26       V27       V28  Amount  Class  
19893  0.631740  0.126366 -0.327633  0.056522  0.033139   29.90    0.0  
19894 -0.390756  0.105615  0.152881 -0.025292  0.021130   16.00    0.0  
19895  0.515660  0.218606 -0.138794  0.026395  0.030885   92.00    0.0  
19896 -1.060827  0.062221  0.150428  0.130266  0.067290    1.99    0.0  
19897       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [10]:
#dataset informatation
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19898 entries, 0 to 19897
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    19898 non-null  int64  
 1   V1      19898 non-null  float64
 2   V2      19898 non-null  float64
 3   V3      19898 non-null  float64
 4   V4      19898 non-null  float64
 5   V5      19898 non-null  float64
 6   V6      19898 non-null  float64
 7   V7      19898 non-null  float64
 8   V8      19898 non-null  float64
 9   V9      19898 non-null  float64
 10  V10     19898 non-null  float64
 11  V11     19897 non-null  float64
 12  V12     19897 non-null  float64
 13  V13     19897 non-null  float64
 14  V14     19897 non-null  float64
 15  V15     19897 non-null  float64
 16  V16     19897 non-null  float64
 17  V17     19897 non-null  float64
 18  V18     19897 non-null  float64
 19  V19     19897 non-null  float64
 20  V20     19897 non-null  float64
 21  V21     19897 non-null  float64
 22

In [12]:
#checking number of missing value in dataset
credit_card_data.isnull().sum()


Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       1
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [13]:
# distribution of legit transaction & fradulent transaction
credit_card_data['Class'].value_counts()

Class
0.0    19812
1.0       85
Name: count, dtype: int64

19812 are legit data transaction whereas 85 are fraud

This Dataset is highly unbalanced

0---> Normal Transaction
1---> Fraudulent Transaction

In [15]:
#seperating the data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [16]:
#printing the shape of two variables or panda series to be precise
print(legit.shape)
print(fraud.shape)

(19812, 31)
(85, 31)


In [17]:
# Statiscal measure of data
legit.Amount.describe()


count    19812.000000
mean        70.169855
std        205.091118
min          0.000000
25%          5.900000
50%         16.070000
75%         59.950000
max       7879.420000
Name: Amount, dtype: float64

In [18]:
fraud.Amount.describe()

count      85.000000
mean       93.869647
std       261.736641
min         0.000000
25%         1.000000
50%         1.000000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [19]:
#compare value for both of these classes i.e fraud and legit by the help of mean
credit_card_data.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    15483.312841 -0.207233  0.215159   0.803126  0.250998 -0.137392   
1.0    17436.164706 -8.862174  6.570214 -12.622087  6.343007 -6.187820   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.104262 -0.108931  0.003328  0.652123  ...  0.036041 -0.050458   
1.0   -2.567646 -8.600077  4.519437 -3.041592  ...  0.725080  0.536833   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.145393 -0.036602  0.011905  0.121909  0.032495  0.010624  0.006940   
1.0   -0.395312 -0.385599 -0.275306  0.359543  0.180456  0.907255  0.093985   

          Amount  
Class             
0.0    70.169855  
1.0    93.869647  

[2 rows x 30 columns]

UNDER SAMPLING


Build a sample dataset containing similar distribution of normal transaction and fraudulent transaction

No of fradulent transaction - 85

In [20]:
legit_sample = legit.sample(n=85)

Concatenating two data frame

In [21]:
new_dataset = pd.concat([legit_sample, fraud] , axis=0 )

#values will be added row wise =0
#values will be added column wise =1

In [22]:
new_dataset.head() #values will be added randomly

Time        V1        V2        V3        V4        V5        V6  \
949      719 -1.313464  1.462090 -0.152403 -1.601424  2.990005  3.359167   
5871    6398 -2.380210 -0.266473  1.090077  0.853215  3.371780  3.790719   
1602    1247  1.055550 -0.282119 -0.074789  0.251727  0.833534  2.307399   
14108  25099 -1.223616 -0.645608  2.402610  0.912546  0.023744 -0.182326   
17777  28908 -0.450114  0.202993  2.195365  0.038529 -0.991969 -0.034317   

             V7        V8        V9  ...       V21       V22       V23  \
949    1.064518 -0.222779  1.392679  ... -0.682337 -0.986381 -0.160630   
5871  -1.123674  0.941963  1.236974  ... -0.306571 -0.514617 -0.443708   
1602  -0.488179  0.699448  0.317465  ... -0.132353 -0.127154 -0.025559   
14108 -0.645850  0.174265 -0.204272  ... -0.192226 -0.231945  0.169684   
17777  0.220404  0.098104  0.457042  ... -0.081643 -0.099554  0.127498   

            V24       V25       V26       V27       V28  Amount  Class  
949    0.914677  0.087702 -0.069327 -0.458951 -0.693665    8.96    0.0  
5871   0.962374  0.088245 -0.280859 -0.635895  0.616903    0.00    0.0  
1602  -2.162523  0.247560  0.452469  0.038747 -0.022087   21.66    0.0  
14108 -0.049261  0.129914 -0.127230  0.059251  0.107628   90.00    0.0  
17777  0.471421 -0.536562  0.975934  0.070054  0.135855   82.60    0.0  

[5 rows x 31 columns]

In [23]:
new_dataset.tail()

Time         V1         V2         V3        V4         V5        V6  \
17480  28755 -30.552380  16.713389 -31.103685  6.534984 -22.105532 -4.977692   
18466  29526   1.102804   2.829168  -3.932870  4.707691   2.937967 -1.800904   
18472  29531  -1.060676   2.608579  -2.971679  4.360089   3.738853 -2.728395   
18773  29753   0.269614   3.549755  -5.810353  5.809370   1.538808 -2.269219   
18809  29785   0.923764   0.344048  -2.880004  1.721680  -3.019565 -0.639736   

              V7         V8        V9  ...       V21       V22       V23  \
17480 -20.371514  20.007208 -3.565738  ...  1.816520 -2.288686 -1.460544   
18466   1.672734  -0.300240 -2.783011  ... -0.106994 -0.250050 -0.521627   
18472   1.987616  -0.357345 -2.757535  ... -0.063168 -0.207385 -0.183261   
18773  -0.824203   0.351070 -3.759059  ...  0.371121 -0.322290 -0.549856   
18809  -3.801325   1.299096  0.864065  ...  0.899931  1.481271  0.725266   

            V24       V25       V26       V27       V28  Amount  Class  
17480  0.183179  2.208209 -0.208824  1.232636  0.356660   99.99    1.0  
18466 -0.448950  1.291646  0.516327  0.009146  0.153318    0.68    1.0  
18472 -0.103679  0.896178  0.407387 -0.130918  0.192177    0.68    1.0  
18773 -0.520629  1.378210  0.564714  0.553255  0.402400    0.68    1.0  
18809  0.176960 -1.815638 -0.536517  0.489035 -0.049729   30.30    1.0  

[5 rows x 31 columns]

In [26]:
new_dataset['Class'].value_counts() #uniformly distributed data

Class
0.0    85
1.0    85
Name: count, dtype: int64

In [27]:
new_dataset.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    15469.988235 -0.411933  0.104753   1.072520  0.281028 -0.104414   
1.0    17436.164706 -8.862174  6.570214 -12.622087  6.343007 -6.187820   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0   -0.017468  0.019046 -0.006275  0.702969  ... -0.115334 -0.133599   
1.0   -2.567646 -8.600077  4.519437 -3.041592  ...  0.725080  0.536833   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.157751 -0.017046  0.063860  0.100851  0.031784  0.031277  0.038321   
1.0   -0.395312 -0.385599 -0.275306  0.359543  0.180456  0.907255  0.093985   

          Amount  
Class             
0.0    66.790235  
1.0    93.869647  

[2 rows x 30 columns]

Spliting the data into features and targets (x and y)

In [28]:
X= new_dataset.drop(columns = 'Class' , axis=1)
Y= new_dataset['Class']

In [29]:
print(X)

        Time         V1         V2         V3        V4         V5        V6  \
949      719  -1.313464   1.462090  -0.152403 -1.601424   2.990005  3.359167   
5871    6398  -2.380210  -0.266473   1.090077  0.853215   3.371780  3.790719   
1602    1247   1.055550  -0.282119  -0.074789  0.251727   0.833534  2.307399   
14108  25099  -1.223616  -0.645608   2.402610  0.912546   0.023744 -0.182326   
17777  28908  -0.450114   0.202993   2.195365  0.038529  -0.991969 -0.034317   
...      ...        ...        ...        ...       ...        ...       ...   
17480  28755 -30.552380  16.713389 -31.103685  6.534984 -22.105532 -4.977692   
18466  29526   1.102804   2.829168  -3.932870  4.707691   2.937967 -1.800904   
18472  29531  -1.060676   2.608579  -2.971679  4.360089   3.738853 -2.728395   
18773  29753   0.269614   3.549755  -5.810353  5.809370   1.538808 -2.269219   
18809  29785   0.923764   0.344048  -2.880004  1.721680  -3.019565 -0.639736   

              V7         V8        V9  

In [30]:
print(Y)

949      0.0
5871     0.0
1602     0.0
14108    0.0
17777    0.0
        ... 
17480    1.0
18466    1.0
18472    1.0
18773    1.0
18809    1.0
Name: Class, Length: 170, dtype: float64


SPLIT THE DATA INTO TRAINING AND TESTING DATA

80- training data
20 - testing data

80% of data in X train

20% data in X test

correspondng of X train will be stored in Y train


similarly of X test wll be stored in Y test


In [33]:
X_train , X_test , Y_train , Y_test = train_test_split(X, Y,test_size=0.2 , stratify=Y,  random_state = 2)

In [34]:
print(X.shape, X_train.shape, X_test.shape)

(170, 30) (136, 30) (34, 30)


MODEL TRAINING
using logistic Regression for Binary classification

In [35]:
#already imported
model = LogisticRegression()


In [41]:
#training the logistic Regression Model with Training Data
model.fit(X_train, Y_train)

LogisticRegression(max_iter=1000, solver='liblinear')

In [39]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=1000)  # Increase max_iter
model.fit(X_train, Y_train)


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000)

In [40]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(solver='liblinear', max_iter=1000)  # Change solver and increase max_iter
model.fit(X_train, Y_train)


LogisticRegression(max_iter=1000, solver='liblinear')

MODEL EVALUATION

 Accuracy Score


Now that our model has learn from the data SO we will only give X train value  and it will predict the value in X train


In [48]:

  X_train_prediction = model.predict(X_train)
  training_data_accuracy = accuracy_score(X_prediction, Y_train)

NameError: name 'X_prediction' is not defined

Accuracy Score on Training Data -

In [47]:
from sklearn.metrics import accuracy_score

# Predict on training data
X_train_prediction = model.predict(X_train)

# Calculate accuracy on training data
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

print(f'Training Data Accuracy: {training_data_accuracy}')


Training Data Accuracy: 0.9926470588235294


In [51]:
# Accuracy Score on Test Data is
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)


In [52]:
print('Accuracy on Test Data :  ', test_data_accuracy)

Accuracy on Test Data :   0.9411764705882353
